community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexMap.java

```diff
@@ -95,32 +130,204 @@ public void foreachIndexProxy( BiConsumer<Long, IndexProxy> consumer )
     {
         Map<K, V> shallowCopy = new HashMap<>( map.size() );
         shallowCopy.putAll( map );
         return shallowCopy;
     }
 
-    public Iterator<LabelSchemaDescriptor> descriptors()
+    private void addDescriptorToLookups( LabelSchemaDescriptor schema )
     {
-        return indexesByDescriptor.keySet().iterator();
+        addToLookup( schema.getLabelId(), schema, descriptorsByLabel );
+
+        for ( int propertyId : schema.getPropertyIds() )
+        {
+            addToLookup( propertyId, schema, descriptorsByProperty );
+        }
     }
 
-    public Iterator<Long> indexIds()
+    private void addToLookup(
+            int key,
+            LabelSchemaDescriptor schema,
+            PrimitiveIntObjectMap<Set<LabelSchemaDescriptor>> lookup )
     {
-        return indexesById.keySet().iterator();
+        Set<LabelSchemaDescriptor> descriptors = lookup.get( key );
+        if ( descriptors == null )
+        {
+            descriptors = new HashSet<>();
+            lookup.put( key, descriptors );
+        }
+        descriptors.add( schema );
     }
 
-    public int size()
+    private void removeFromLookup(
+            int key,
+            LabelSchemaDescriptor schema,
+            PrimitiveIntObjectMap<Set<LabelSchemaDescriptor>> lookup )
     {
-        return indexesById.size();
+        Set<LabelSchemaDescriptor> descriptors = lookup.get( key );
+        descriptors.remove( schema );
+        if ( descriptors.isEmpty() )
+        {
+            lookup.remove( key );
+        }
+    }
+
+    private static Map<LabelSchemaDescriptor, IndexProxy> indexesByDescriptor( Map<Long,IndexProxy> indexesById )
+    {
+        Map<LabelSchemaDescriptor, IndexProxy> map = new HashMap<>();
+        for ( IndexProxy proxy : indexesById.values() )
+        {
+            map.put( proxy.schema(), proxy );
+        }
+        return map;
+    }
+
+    private static Map<LabelSchemaDescriptor, Long> indexIdsByDescriptor( Map<Long,IndexProxy> indexesById )
+    {
+        Map<LabelSchemaDescriptor, Long> map = new HashMap<>();
+        for ( Map.Entry<Long,IndexProxy> entry : indexesById.entrySet() )
+        {
+            map.put( entry.getValue().schema(), entry.getKey() );
+        }
+        return map;
+    }
+
+    /**
+     * Get descriptors affected by changed properties. Implementation is smart about whether doing
+     * the lookup using the unchanged labels or the changed properties given the smallest final
+     * set of indexes.
+     *
+     * @param unchangedLabels set of labels that are unchanged
+     * @param properties set of properties that have changed
+     * @return set of LabelSchemaDescriptors describing the potentially affected indexes
+     */
+    private Set<LabelSchemaDescriptor> getDescriptorsByProperties(
+            long[] unchangedLabels,
+            PrimitiveIntCollection properties )
+    {
+        int nIndexesForLabels = countIndexesByLabels( unchangedLabels );
```

Comment:
Yeah, but then we need to first collect the subparts into a list for both the label and property paths, before turning the smaller one into a set and return. This means some extra construction work. 

And if we do not want to construct objects when there are no affected indexes, we need ´null´-checks which makes stuff harder to read.